### ISSUE TITLE
BUG: `pz.chk.check_structure` erroneously requires all named array axes to be strings.

### ISSUE THREAD

**Issue by amifalk** [google-deepmind/penzai#132]
```python
import jax
from penzai import pz

my_layer = pz.nn.Linear.from_config(
    "my_layer", jax.random.key(0), input_axes={"in_axis": 2}, output_axes={"out_axis": 3}
)

nonstr_batch_axis = pz.nx.TmpPosAxisMarker()

my_layer(pz.nx.zeros({"in_axis": 2, nonstr_batch_axis: 3}))
```

The following raises an assertion error here: https://github.com/google-deepmind/penzai/blob/aac7808a3d1269ea9885094d78d2274d56fc7449/penzai/core/shapecheck.py#L621

I don't think any assumptions of shapecheck require this, and it limits the ability to do automatic axis naming without worry of axis name overlaps.

**Comment by simbot912** [comment]
This is a useful bug report! The `pz.chk.check_structure` assertion at line 621 seems overly restrictive — requiring all named array axes to be strings when `TmpPosAxisMarker()` is a legitimate way to create temporary batch axes.

The issue is that `TmpPosAxisMarker` instances are intentionally non-string axis markers for temporary computations, and the shapecheck shouldn't reject them. A potential fix would be to modify the assertion to allow both string axis names and `TmpPosAxisMarker` instances:

```python
# Instead of requiring isinstance(axis_name, str), check for either
assert isinstance(axis_name, (str, TmpPosAxisMarker))
```

Would you be open to submitting a PR with this fix plus a test case?
